### PR TITLE
GARGOYLE: New Gargoyle (GLK) engine

### PR DESCRIPTION
**Pull request by dreammaster**
It's been a long time coming, but this engine aims to close the major remaining gap of story-based adventures that ScummVM doesn't yet support.. that of interactive fiction (aka "text adventure"). Some of them, like Infocom, rivaled Lucasarts and Sierra for making compelling stories. Some of the later Infocom ones, in particular, are hybrid text and graphics, making them close to point and click adventures. Unlike my earlier Gargoyle engine of nearly a decade ago, which was a specific implementation of just the Frotz interpreter (for Infocom games), this new implementation uses the GLK portability layer, that was developed some years back, to ease porting the various IF interpreters to different systems. As such, it will make it easier to add support for the different interpreters as sub-engines within this engine. 

This new version also has several further advantages over my previous attempt. The GLK layer provides  support for Unicode, which allows more flexible display of foreign languages.. coupled with ScummVM's support for U32String and passing them to the font drawing routines. ScummVM also supports copying and pasting to/from the clipboard, which wasn't available back when I did my original engine.

At the moment, I've only implemented the Scott Adams interpreter as a sub-engine as a proof of concept. It should be fairly easy to likewise add other interpreters in afterwards. I've kept all the method names and types available as in the original GLK, so the bulk of work for new sub-engines will simply be setting them up in their own sub-namespace, massaging them as necessary to have a proper class structure, and any miscellaneous ScummVM things like the detection code, and supporting loading from the launcher.

The engine comes with it's own data file, which is the set of freely available GLK TTF fonts packaged up in a simple container format. The Scott Adams games are freely available from his website at http://www.msadams.com/. The scott sub-engine detects the games using md5 matches on any filename with either a .dat or .saga extension. So you can rename the games as you please for better naming clarity. 

Currently, I've implemented the bulk of the GLK compatibility layer, except for the sound handling and some of the graphic methods. These are waiting on implementing a sub-engine which makes use of them, like Frotz, so I can test the results of what I implement.

### CODE REVIEW

File: engines/gargoyle/scott/scott.cpp [hunk 1]
```diff
@@ -0,0 +1,1263 @@
...
+	va_list va;
+	va_start(va, count);
+	unsigned char c = f->readByte();
+
+	for (size_t idx = 0; idx < count; ++idx) {
+		while (f->pos() < f->size() && Common::isSpace(c))
+			c = f->readByte();
+
+		// Get the next value
+		int *val = (int *)va_arg(va, int);
```
Comment (bgK) [inline engines/gargoyle/scott/scott.cpp]: Shouldn't this be:
```suggestion
		int *val = va_arg(va, int *);
```

File: engines/gargoyle/scott/detection.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+struct ScottGame {
+	const char *_md5;
+	const char *_gameId;
+	int32 _filesize;
+	const char *_desc;
+};
+
+const ScottGame SCOTT_GAMES[] = {
+	{ "ae541fc1085da2f7d561b72ed20a6bc1", "adventureland", 18003, "Adventureland" },
```
Comment (bgK) [inline engines/gargoyle/scott/detection.cpp]: I have downloaded the datafiles from the author's site (http://www.msadams.com/downloads/advgames.zip, more precisely). However the md5s don't seem to match. Am I missing something?

File: engines/gargoyle/scott/detection.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	{ nullptr, nullptr, 0, nullptr }
+};
+
+bool ScottMetaEngine::detectGames(const Common::FSList &fslist, DetectedGames &gameList) {
+	Common::File gameFile;
+	Common::String md5;
+
+	// Loop through the files of the folder
+	for (Common::FSList::const_iterator file = fslist.begin(); file != fslist.end(); ++file) {
+		if (file->isDirectory() || !file->getName().hasSuffix(".saga"))
```
Comment (bgK) [inline engines/gargoyle/scott/detection.cpp]: Maybe the `.dat` extension should be accepted as well since that's what was originally used.

**Comment by bgK** [comment]
How do you think this engine should position itself with regard to the upstream Gargoyle project? Is it meant to replace it? Was Gargoyle used as a starting point to introduce a GLK implementation in ScummVM? If it is the latter then maybe the engine should be renamed to make it clear the two are separate, and not to unintentionally harm the Gargoyle project.

File: engines/gargoyle/scott/detection.cpp (hunk 3 above)
Comment (dreammaster) [reply engines/gargoyle/scott/detection.cpp]: Whoops. That was the way I intended to set up the code, which is why the adventure list doesn't have explicit filenames. I'll fix that.

Comment (dreammaster) [reply engines/gargoyle/scott/detection.cpp]: Interesting.. turns out I had accidentally originally downloaded the PDA version, and got the data files from it, renaming the data files to .saga for clarity. But what's particularly interesting is that the PDA zip has an extra game "Buckaroo Banzai", whilst the Windows zip archive doesn't have it, but has Marvel Adventure #1 and a Sampler. And all of the game files in that have different sizes & MD5s to the PDA version. So cool, extra games. I'm about to commit in extra detection entries for them.

**Comment by dreammaster** [comment]
> How do you think this engine should position itself with regard to the upstream Gargoyle project? Is it meant to replace it? Was Gargoyle used as a starting point to introduce a GLK implementation in ScummVM? If it is the latter then maybe the engine should be renamed to make it clear the two are separate, and not to unintentionally harm the Gargoyle project.

That's an interesting question. I really haven't been active in the Interactive Fiction community for over a decade, so I'm not really in a position to make a comment on whether it would cause confusion or not. My gut feeling though is that there won't be a problem.. engine names aren't really "in the face" of end users, so it shouldn't matter too much what it's called. But I'll raise the question on the Garglk Google group, just to be on the safe side.

**Comment by ccawley2011** [comment]
>The engine comes with it's own data file, which is the set of freely available GLK TTF fonts packaged up in a simple container format.

Considering that gargoyle.dat is simply packaging several other files, it might be better to use a .zip archive instead of creating a new container format. A couple of other engines (namely Wintermute and MacVenture) already use zip files for this purpose.

In addition, it might be a good idea to rename it to "fonts.dat" or something similarly generic, so it can be used by other engines (Z-Vision, for example).

**Comment by dreammaster** [comment]
> Considering that gargoyle.dat is simply packaging several other files, it might be better to use a .zip archive instead of creating a new container format. A couple of other engines (namely Wintermute and MacVenture) already use zip files for this purpose.
> 
> In addition, it might be a good idea to rename it to "fonts.dat" or something similarly generic, so it can be used by other engines (Z-Vision, for example).

Good idea, even if it does make the entire create_gargoyle tool redundant ;)

File: engines/gargoyle/detection.cpp [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+
+	return detectedGames;
+}
+
+static Gargoyle::GargoyleGameDescription gameDescription;
+
+ADDetectedGames GargoyleMetaEngine::detectGame(const Common::FSNode &parent, const FileMap &allFiles, Common::Language language, Common::Platform platform, const Common::String &extra) const {
+	static char gameId[100];
+	strcpy(gameId, ConfMan.get("gameid").c_str());
+	Common::String filename = ConfMan.get("filename");
```
Comment (bgK) [inline engines/gargoyle/detection.cpp]: I'm not so sure about storing the filename in the configuration file when detecting the game and then using it when launching. That does not work with the command line interface where you can start a game just from a gameid and a path, without a target / bypassing the detection code (`scummvm --path=~/games/IF/Scott/ adventureland`).

File: engines/gargoyle/detection.cpp [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+
+#include "common/config-manager.h"
+#include "common/file.h"
+#include "gargoyle/detection_tables.h"
+#include "gargoyle/frotz/detection.h"
+#include "gargoyle/frotz/frotz.h"
+#include "gargoyle/scott/detection.h"
+#include "gargoyle/scott/scott.h"
+
+class GargoyleMetaEngine : public AdvancedMetaEngine {
```
Comment (bgK) [inline engines/gargoyle/detection.cpp]: IMO, this class should either extend `AdvancedMetaEngine` and use its standard detection code or extend `MetaEngine` with the current detection code.

File: engines/gargoyle/detection.cpp (hunk 4 above)
Comment (dreammaster) [reply engines/gargoyle/detection.cpp]: That may indeed be a problem. But I'm not sure if it's avoidable. Currently, adventures only tend to have a single adventure per folder, with a known set of files. Or, in rare cases like DOTT, technically two games, but still with a distinct set of known files. For the case of Interactive Fiction, it's more problematic, because dozens of games may end up being in the same folder, including multiple versions of the same game.

For example, I have a folder structure with a folder for each game, and within each all the different versions of that game (or at least, those that I've collected). If you look at https://github.com/dreammaster/scummvm/blob/gargoyle/engines/gargoyle/frotz/detection_tables.cpp  you'll see that I have 5 different versions of The Hitchhiker's Guide to the Galaxy. If I used the command line like in your example, scummvm --path=~/games/if/hhgttg/ hhgttg   how would it know which version of the game to launch. It gets even worse with popular interactive fiction formats like zcode, where there are hundreds, maybe thousands, of fan games which could all be present in the one folder.

I'm not really familiar with the detection code too much. In all the cases in the past, it was easy simple to use AdvancedMetaEngine with a hardcoded set of files/sizes/md5s, so I didn't really look into how it did it's magic. :). I don't suppose the path can optionally specify a specific filename as well? Spitballing ideas.. even if it doesn't already, maybe it would be something better to do as a change to the advanced meta engine.. in that when a path specifies a specific filename, it would transparently treat it as if in the parent directory with only that one file passed to the detectGames method. That way, I wouldn't need a separate filename field. It still doesn't really do away with the messy startup code I currently need for the startup, though.

File: engines/gargoyle/detection.cpp (hunk 5 above)
Comment (dreammaster) [reply engines/gargoyle/detection.cpp]: I'll have a look into extending from MetaEngine instead then.

**Comment by dreammaster** [comment]
I got a response to my query regarding the name, and the preference was to have the name different. "GarGLK" is used as a description for Gargoyle's port of the GLK spec, so "ScummGLK" was suggested. I'll rebase the history to change the commit messages accordingly, and then commit in a change to the namespace.

**Comment by digitall** [comment]
@dreammaster : I know that this is the ScummVM port of GLK, hence why ScummGLK, but maybe a more generic engine name would be better i.e. interactivefiction or textadventure ?

**Comment by dreammaster** [comment]
@digitall I disagree, those names seem too wordy, in my opinion. Also, we normally try to name engines by internal names where possible. Since it's core is Glk, I've now changed the core engine folder/namespace to Glk, though the visual nane in the about will read 'ScummGlk'. I'm just in the middle of doing that right now.

**Comment by digitall** [comment]
@dreammaster : Hmm... Okay. I see your point. Having the internal folder/namespace as glk is better than scummglk as confusing with the SCUMM engine... and as you said the visual name is ScummVM implementation of GLK thus ScummGlk  ... Okay, seems reasonable. If we do have any text adventures in future which do not fit under Glk, they would probably be a separate engine anyway. Sorry to be a bother, but thought I should play "Devil's Advocate" before the name is cast in stone ... or at least git history :)

**Comment by dreammaster** [comment]
Indeed.. II still have hopes of eventually implementing the earlier Legend text/graphic adventures, they'll all go in the preliminary legend engine I was working on previously.

**Comment by dreammaster** [comment]
Closing this in favor of creating a pull request for a new branch where I've rebased the historical commit messages.

**Comment by digitall** [comment]
For traceability, the new pull request is PR #1393
